atlasdb-coordination-impl/src/main/java/com/palantir/atlasdb/coordination/CoordinationService.java

```diff
@@ -0,0 +1,64 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.coordination;
+
+import java.util.Optional;
+import java.util.function.Function;
+
+import com.palantir.atlasdb.keyvalue.impl.CheckAndSetResult;
+
+/**
+ * A {@link CoordinationService} is used to agree on values being relevant or correct at a given timestamp.
+ * The sequence of values being agreed should evolve in a backwards consistent manner; that is, if I read a value
+ * for some timestamp TS and make a decision at TS, all future values written to the {@link CoordinationService} should
+ * also ensure that no decision that would be inconsistent with the decision made at TS would be made at TS. In general,
+ * this means that one should avoid changes that affect behaviour of values below the validity bound.
+ *
+ * It is the responsibility of the caller to preserve the above property.
+ */
+public interface CoordinationService<T> {
+    /**
+     * Returns a value that the coordination service has agreed is appropriate at the provided timestamp.
+     *
+     * The value returned by this method may change over time. However, values returned by this method would
+     * be consistent at this timestamp - that is, the service should never return values for the same timestamp
+     * that would be contradictory or result in different operation for uses at this timestamp.
+     *
+     * This operation will return an empty Optional if no value has been agreed upon yet, or if the current value agreed
+     * by the coordination service is not valid at the provided timestamp.
+     *
+     * @param timestamp timestamp to retrieve the coordinated value for
+     * @return value associated with that timestamp
+     */
+    Optional<ValueAndBound<T>> getValueForTimestamp(long timestamp);
+
+    /**
+     * Updates the value that this coordination service stores, marking this value as valid up to a certain timestamp.
+     * It may transpire that the value in the coordination service is already acceptable - if so, the provided
+     * transformation should simply return the original input. If the transformation returns a value with a bound
+     * less than or equal to the bound of the input, the results of the transformation will not be stored.
+     *
+     * Evolutions of the value must be compatible in terms of backwards consistency as defined in the class docs.
+     *
+     * The {@link ValueAndBound} returned by the transform must contain a value.
+     *
+     * @param transform transformation to apply to the existing value and bound the coordination service agrees on
+     * @return a {@link CheckAndSetResult} indicating whether the transform was applied and the current value
+     */
+    CheckAndSetResult<ValueAndBound<T>> tryTransformCurrentValue(
```

Comment:
Would it be better if the bounds management was handled internally? Clients would propose behavior, and internally we would make it apply at the next reasonable opportunity (= some future timestamp).

Reply:
(offline discussion: I'll attempt to do this. This still needs to be a function, but it will probably be enough to have this be a transform on just the value, rather than including the bounds)